leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/senml/LwM2mNodeSenMLJsonDecoder.java

```diff
@@ -0,0 +1,327 @@
+            LwM2mPath bnPath = new LwM2mPath(baseName);
+
+            // check returned base name path is under requested path
+            if (requestPath.getObjectId() != null && bnPath.getObjectId() != null) {
+                if (!bnPath.getObjectId().equals(requestPath.getObjectId())) {
+                    throw new CodecException("Basename path [%s] does not match requested path [%s].", bnPath,
+                            requestPath);
+                }
+                if (requestPath.getObjectInstanceId() != null && bnPath.getObjectInstanceId() != null) {
+                    if (!bnPath.getObjectInstanceId().equals(requestPath.getObjectInstanceId())) {
+                        throw new CodecException("Basename path [%s] does not match requested path [%s].", bnPath,
+                                requestPath);
+                    }
+                    if (requestPath.getResourceId() != null && bnPath.getResourceId() != null) {
+                        if (!bnPath.getResourceId().equals(requestPath.getResourceId())) {
+                            throw new CodecException("Basename path [%s] does not match requested path [%s].", bnPath,
+                                    requestPath);
+                        }
+                    }
+                }
+            }
+            return bnPath;
+        }
+        return null;
+    }
+
+    /**
+     * Group all SenML record by instanceId
+     * 
+     * @param records
+     * @param baseName
+     *
+     * @return a map (instanceId => collection of SenML Record)
+     */
+    private static Map<Integer, Collection<SenMLRecord>> groupRecordsByInstanceId(Collection<SenMLRecord> records,
+            LwM2mPath baseName) throws CodecException {
+        Map<Integer, Collection<SenMLRecord>> result = new HashMap<>();
+
+        for (SenMLRecord record : records) {
+            // Build resource path
+            LwM2mPath nodePath = record.getName() == null ? baseName : baseName.append(record.getName());
+
+            // Validate path
+            if (!nodePath.isResourceInstance() && !nodePath.isResource()) {
+                throw new CodecException(
+                        "Invalid path [%s] for resource, it should be a resource or a resource instance path",
+                        nodePath);
+            }
+
+            // Get SenML records for this instance
+            Collection<SenMLRecord> recordForInstance = result.get(nodePath.getObjectInstanceId());
+            if (recordForInstance == null) {
+                recordForInstance = new ArrayList<>();
+                result.put(nodePath.getObjectInstanceId(), recordForInstance);
+            }
+
+            // Add it to the list
+            recordForInstance.add(record);
+        }
+
+        // Create an entry for an empty instance if possible
+        if (result.isEmpty() && baseName.getObjectInstanceId() != null) {
+            result.put(baseName.getObjectInstanceId(), new ArrayList<SenMLRecord>());
+        }
```

Comment:
If there is no more baseName parameter, we will probably remove this part of the code.
And so we should handle the case where result if empy in calling method.